pkg/deploy/controller/deployment/factory.go

```diff
@@ -1,177 +1,203 @@
+	})
+
+	c.rcStoreSynced = rcInformer.HasSynced
+	c.podStoreSynced = podInformer.HasSynced
+
+	return c
 }
 
-// makeContainer creates containers in the following way:
-//
-//   1. For the Recreate and Rolling strategies, strategy, use the factory's
-//      DeployerImage as the container image, and the factory's Environment
-//      as the container environment.
-//   2. For all Custom strategies, or if the CustomParams field is set, use
-//      the strategy's image for the container image, and use the combination
-//      of the factory's Environment and the strategy's environment as the
-//      container environment.
-//
-func (factory *DeploymentControllerFactory) makeContainer(strategy *deployapi.DeploymentStrategy) *kapi.Container {
-	image := factory.DeployerImage
-	var environment []kapi.EnvVar
-	var command []string
-
-	set := sets.NewString()
-	// Use user-defined values from the strategy input.
-	if p := strategy.CustomParams; p != nil {
-		if len(p.Image) > 0 {
-			image = p.Image
-		}
-		if len(p.Command) > 0 {
-			command = p.Command
-		}
-		for _, env := range strategy.CustomParams.Environment {
-			set.Insert(env.Name)
-			environment = append(environment, env)
+// Run begins watching and syncing.
+func (c *DeploymentController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Wait for the dc store to sync before starting any work in this controller.
+	ready := make(chan struct{})
+	go c.waitForSyncedStores(ready, stopCh)
+	select {
+	case <-ready:
+	case <-stopCh:
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down deployment controller")
+	c.queue.ShutDown()
+}
+
+func (c *DeploymentController) waitForSyncedStores(ready chan<- struct{}, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	for !c.rcStoreSynced() || !c.podStoreSynced() {
+		glog.V(4).Infof("Waiting for the rc and pod caches to sync before starting the deployment controller workers")
+		select {
+		case <-time.After(StoreSyncedPollPeriod):
+		case <-stopCh:
+			return
 		}
 	}
+	close(ready)
+}
+
+func (c *DeploymentController) addReplicationController(obj interface{}) {
+	rc := obj.(*kapi.ReplicationController)
+	// Filter out all unrelated replication controllers.
+	if !deployutil.IsOwnedByConfig(rc) {
+		return
+	}
+
+	c.enqueueReplicationController(rc)
+}
+
+func (c *DeploymentController) updateReplicationController(old, cur interface{}) {
+	rc := cur.(*kapi.ReplicationController)
+	// Filter out all unrelated replication controllers.
+	if !deployutil.IsOwnedByConfig(rc) {
+		return
+	}
+
+	c.enqueueReplicationController(rc)
+}
+
+func (c *DeploymentController) updatePod(old, cur interface{}) {
+	// A periodic relist will send update events for all known controllers.
+	if kapi.Semantic.DeepEqual(old, cur) {
+		return
+	}
+
+	pod := cur.(*kapi.Pod)
+
+	if rc, err := c.rcForDeployerPod(pod); err == nil && rc != nil {
```

Comment:
note that latent RCs disturbing the return are ok because that means that the RC will be bumped again.